ecosystem/indexer-grpc/indexer-grpc-utils/src/in_memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+// Copyright © Aptos Foundation
+
+use crate::compression_util::{CacheEntry, StorageFormat};
+use anyhow::Context;
+use aptos_protos::transaction::v1::Transaction;
+use dashmap::DashMap;
+use itertools::Itertools;
+use prost::Message;
+use redis::AsyncCommands;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+// Internal lookup retry interval for in-memory cache.
+const IN_MEMORY_CACHE_LOOKUP_RETRY_INTERVAL_MS: u64 = 10;
+// Max cache size in bytes: 5 GB.
+const MAX_IN_MEMORY_CACHE_CAPACITY_IN_BYTES: u64 = 5_000_000_000;
+// Max cache entry TTL: 30 seconds.
+// const MAX_IN_MEMORY_CACHE_ENTRY_TTL: u64 = 30;
+// Warm-up cache entries. Pre-fetch the cache entries to warm up the cache.
+const WARM_UP_CACHE_ENTRIES: u64 = 20_000;
+const MAX_REDIS_FETCH_BATCH_SIZE: usize = 500;
+
+#[derive(Debug, Clone, Copy)]
+struct CacheMetadata {
+    total_size_in_bytes: u64,
+    latest_version: u64,
+    first_version: u64,
+}
+
+/// InMemoryCache is a simple in-memory cache that stores the protobuf Transaction.
+pub struct InMemoryCache {
+    /// Cache maps the cache key to the deserialized Transaction.
+    cache: Arc<DashMap<u64, Arc<Transaction>>>,
```

Comment:
Why do you need both Arc here?

Reply:
first one is to allow cache to be passed around threads(read, write).

second one is to avoid copying data while holding the internal lock.

Reply:
Isn't DashMap already Send + Sync?

Reply:
hmm true; read it in wrong way 🤦 . the first `Arc` can go away.

Reply:
done! good catch.

Reply:
Does &DashMap work or is it too hard to annotate the lifetime?

Reply:
dumb question warning lol: i feel this is a common pattern, how to pass Dashmap around? 

the pattern in short is

```rust

let cache = dashmap::new();

spawn_task_one(cache) {
   tokio::spawn()
}


spawn_task_two(cache) {
   tokio::spawn()
}

// ... program exits
drop(dashmap)
```

My understanding is `static` lifetime is required if we want to pass around the &Dashmap, right?

Reply:
i'll start the conversation in the follow-up PR and experiment tokio_scope.

at the same time, i am going to merge haha